src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using System.Runtime.Serialization;
+using System.Runtime.Serialization.Json;
+using System.Text;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using ResourceHashesByNameDictionary = System.Collections.Generic.Dictionary<string, string>;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Build
+{
+    public class MarkLazyLoadDependencies : Task
+    {
+        [Required]
+        public ITaskItem[] AssemblyReferences { get; set; }
+
+        [Required]
+        public ITaskItem[] PackageReferences { get; set; }
+
+        [Required]
+        public ITaskItem[] ExplicitlazyLoad { get; set; }
+
+        [Output]
+        public ITaskItem[] LazyLoadedAssemblies { get; set; }
+
+        private List<string> _visitedAssemblies = new List<string>();
+        private HashSet<ITaskItem> _lazyLoadAssemblies = new HashSet<ITaskItem>();
+
+        private const string LAZY_LOAD_LABEL = "BlazorLazyLoad";
+        private const string REFERENCE_SOURCE_KEY = "ReferenceSourceTarget";
+        private const string PROJECT_REFERENCE = "ProjectReference";
+        private const string ASSEMBLY_REFERENCE = "ResolveAssemblyReference";
+
+        public override bool Execute()
+        {
+            // Add assemblies that have been explicitly set by the user in a
+            // `BlazorLazyLoad` item group.
+            //
+            // <ItemGroup>
+            //  <BlazorLazyLoad Include="SomeBigAssembly.dll" />
+            // </ItemGroup>
+            _lazyLoadAssemblies.UnionWith(ExplicitlazyLoad);
+
+            foreach (var assembly in AssemblyReferences)
+            {
+                try
+                {
+                    ProcessAssembly(assembly);
+                }
+                catch (Exception e)
+                {
+                    Log.LogErrorFromException(e);
+                }
+            }
+
+            LazyLoadedAssemblies = _lazyLoadAssemblies.ToArray();
+
+            return !Log.HasLoggedErrors;
+        }
+
+        private void ProcessAssembly(ITaskItem assembly)
+        {
+            var fileName = assembly.GetMetadata("FileName");
+
+            var visited = _visitedAssemblies.Contains(fileName);
+            if (visited)
```

Comment:
You could replace `! .Contains` + `.Add` with

```suggestion
            if (!_visitedAssemblies.Add(fileName))
```

Reply:
Ended up changing the type of `_visitedAssemblies` to `HashSet` to use this API.